### PR TITLE
Added Sequential Gauss-Newton optimizer

### PR DESCRIPTION
**Pull request by bayardj**
In order to perform sequential batch least square estimaion, it could be interesting to implement a sequential Gauss Newton optimizer.

### CODE REVIEW

File: hipparchus-optim/src/main/java/org/hipparchus/optim/LocalizedOptimFormats.java [hunk 1]
```diff
@@ -57,7 +57,8 @@
     TRUST_REGION_STEP_FAILED("trust region step has failed to reduce Q"),
     UNABLE_TO_PERFORM_QR_DECOMPOSITION_ON_JACOBIAN("unable to perform Q.R decomposition on the {0}x{1} jacobian matrix"),
     UNABLE_TO_SOLVE_SINGULAR_PROBLEM("unable to solve: singular problem"),
-    UNBOUNDED_SOLUTION("unbounded solution");
+    UNBOUNDED_SOLUTION("unbounded solution"),
+    UNEQUAL_VECTOR_DIMENSIONS("unequal vector dimensions : {0} and {1}");
```
Comment (maisonobe) [inline hipparchus-optim/src/main/java/org/hipparchus/optim/LocalizedOptimFormats.java:61]: There is already an DIMENSIONS_MISMATCH constant

File: hipparchus-optim/src/main/java/org/hipparchus/optim/nonlinear/vector/leastsquares/SequentialGaussNewtonOptimizer.java [hunk 2]
```diff
@@ -0,0 +1,340 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
```
Comment (maisonobe) [inline hipparchus-optim/src/main/java/org/hipparchus/optim/nonlinear/vector/leastsquares/SequentialGaussNewtonOptimizer.java:2]: For new files, the header should refer to Hipparchus, not Apache.
See for example the files added recently in org.hipparchus.special.jacobi

File: hipparchus-optim/src/main/resources/assets/org/hipparchus/optim/LocalizedOptimFormats_fr.utf8 [hunk 3]
```diff
@@ -27,3 +27,4 @@ TRUST_REGION_STEP_FAILED = l''étape de la région de confiance n''a pas réduit
 UNABLE_TO_PERFORM_QR_DECOMPOSITION_ON_JACOBIAN = impossible de calculer la factorisation Q.R de la matrice jacobienne {0}x{1}
 UNABLE_TO_SOLVE_SINGULAR_PROBLEM = résolution impossible : problème singulier
 UNBOUNDED_SOLUTION = solution non bornée
+UNEQUAL_VECTOR_DIMENSIONS = dimensions de vecteur non égales : {0} and {1}
```
Comment (maisonobe) [inline hipparchus-optim/src/main/resources/assets/org/hipparchus/optim/LocalizedOptimFormats_fr.utf8:30]: ditto about DIMENSIONS_MISMATCH

File: hipparchus-optim/src/test/java/org/hipparchus/optim/nonlinear/vector/leastsquares/AbstractSequentialLeastSquaresOptimizerAbstractTest.java [hunk 4]
```diff
@@ -0,0 +1,566 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
```
Comment (maisonobe) [inline hipparchus-optim/src/test/java/org/hipparchus/optim/nonlinear/vector/leastsquares/AbstractSequentialLeastSquaresOptimizerAbstractTest.java:2]: ditto about header on new files

File: hipparchus-optim/src/test/java/org/hipparchus/optim/nonlinear/vector/leastsquares/SequentialGaussNewtonOptimizerWithCholeskyTest.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
```
Comment (maisonobe) [inline hipparchus-optim/src/test/java/org/hipparchus/optim/nonlinear/vector/leastsquares/SequentialGaussNewtonOptimizerWithCholeskyTest.java:2]: ditto about headers on new files

File: hipparchus-optim/src/test/java/org/hipparchus/optim/nonlinear/vector/leastsquares/SequentialGaussNewtonOptimizerWithLUTest.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
```
Comment (maisonobe) [inline hipparchus-optim/src/test/java/org/hipparchus/optim/nonlinear/vector/leastsquares/SequentialGaussNewtonOptimizerWithLUTest.java:2]: ditto about header on new files

File: hipparchus-optim/src/test/java/org/hipparchus/optim/nonlinear/vector/leastsquares/SequentialGaussNewtonOptimizerWithQRNormalTest.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
```
Comment (maisonobe) [inline hipparchus-optim/src/test/java/org/hipparchus/optim/nonlinear/vector/leastsquares/SequentialGaussNewtonOptimizerWithQRNormalTest.java:2]: ditto about header on new files

File: hipparchus-optim/src/test/java/org/hipparchus/optim/nonlinear/vector/leastsquares/SequentialGaussNewtonOptimizerWithQRTest.java [hunk 8]
```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
```
Comment (maisonobe) [inline hipparchus-optim/src/test/java/org/hipparchus/optim/nonlinear/vector/leastsquares/SequentialGaussNewtonOptimizerWithQRTest.java:2]: ditto about header on new files

File: hipparchus-optim/src/test/java/org/hipparchus/optim/nonlinear/vector/leastsquares/SequentialGaussNewtonOptimizerWithSVDTest.java [hunk 9]
```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
```
Comment (maisonobe) [inline hipparchus-optim/src/test/java/org/hipparchus/optim/nonlinear/vector/leastsquares/SequentialGaussNewtonOptimizerWithSVDTest.java:2]: ditto about header on new files

File: hipparchus-optim/src/main/java/org/hipparchus/optim/nonlinear/vector/leastsquares/SequentialGaussNewtonOptimizer.java [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+    public Evaluation getOldEvaluation() {
+        return this.oldEvaluation;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Optimum optimize(final LeastSquaresProblem lsp) {
+        // create local evaluation and iteration counts
+        final Incrementor evaluationCounter = lsp.getEvaluationCounter();
+        final Incrementor iterationCounter = lsp.getIterationCounter();
```
Comment (maisonobe) [inline hipparchus-optim/src/main/java/org/hipparchus/optim/nonlinear/vector/leastsquares/SequentialGaussNewtonOptimizer.java:126]: Are two different counters needed here? They seem to be incremented in parallel.

**Review by maisonobe — COMMENTED** [review]
Very good, thanks for the contribution.
My comments are very minor ones.

**Comment by wardev** [comment]
Hi Bayardj,

Thanks for the contribution. I've also seen this called Bayesian least squares as it incorporates an a priori estimate and covariance into the least squares solution. 

It would be nice to include in the javadoc how a user with a state estimate and a state covariance matrix would pass those to your class. That would clarify that it solves that problem as well. It would be good to include this in this PR.

It does make me wish the interface for LSP allowed accumulating the normal equations a measurement at a time as a memory optimization since you always form the normal matrix in this algorithm.

You could also provide an option for not forming the normal equations by appending the Jacobians and residuals into a larger matrix. That could let the user make the trade off on speed vs numerical stability and perhaps help with some of the failures you're seeing in the test cases.

Regards,
Evan

**Comment by BryanCazabonne** [comment]
Hi Evan and Luc, 

For information Julie is my intern. The motivation of this PR is to have a sequential least-squares optimizer in order to perform orbit determination using a "sequential" batch least squares algorithme (See; Chapter 10.5 of _Fundamentals of Astrodynamics_).

> It would be nice to include in the javadoc how a user with a state estimate and a state covariance matrix would pass those to your class. That would clarify that it solves that problem as well. It would be good to include this in this PR

+1 for adding this information in the Javadoc. It could be very useful for users.

> You could also provide an option for not forming the normal equations by appending the Jacobians and residuals into a larger matrix. That could let the user make the trade off on speed vs numerical stability and perhaps help with some of the failures you're seeing in the test cases.

That is a very interesting aspect that we would like to add later. Indeed, the second main task of Julie's internship is the implementation of an [Extended Semi-analytical Kalman Filter](https://ui.adsabs.harvard.edu/abs/1986JAnSc..34..147W/abstract) in Orekit. We have an important time constraint on this second point. Indeed, we submited an abstract for the next [Astrodynamics Specialist Conference](https://www.space-flight.org/docs/2021_summer/2021_summer.html) and we will need results before the end of July for the paper submission (we keep our fingers crossed that the abstract will be accepted :smile:).
Therefore, we planed to implement the option about normal equations just after the conference.

Is it ok for you or do you think we have to add the option before merging the branch into `master`?

Best regards,
Bryan

**Comment by wardev** [comment]
Hi Bryan, Julie,

Welcome Julie!

I think it is ready to merge once the javadoc is updated so that users will understand the nice capability that has been implemented here. I'm looking at Section 4.4.2 of Tabley, Schutz, and Born's _Statistical Orbit Determination_ and Julie has created a straight forward implementation of it.

Those bigger comments were me thinking about nice features to have in the future, not things we need to implement now. I should have been more clear.

I would be interested in reading your paper when you're ready to share it. It is a shame that the conference is virtual this year, but it does make travel arrangements easier. :)

Regards,
Evan

**Comment by BryanCazabonne** [comment]
>   It would be nice to include in the javadoc how a user with a state estimate and a state covariance matrix would pass those to your class. That would clarify that it solves that problem as well. It would be good to include this in this PR

   > +1 for adding this information in the Javadoc. It could be very useful for users.

I just realized something. With the current implementation of the `SequentialGaussNewtonOptimizer` class, it will not be possible for a user to use the class without specifying the previous residual vector and the previous Jacobian matrix.

Adding the ability to perform sequentiel batch least squares from the previous estimated state and the corresponding previous covariance matrix  **only** state is also something we would like to implement in the future. This is very useful in the case of satellite operations.  However, as I mentioned before, we have not implemented this yet due to our time constraints for the Astrodynamics Specialist Conference...

I will suggest to Julie to add in the Javadoc how to use the `SequentialGaussNewtonOptimizer` class with the previous estimated state, the Jacobian matrix and the measurement residuals. This is a first interesting help in the Javadoc.

What do you think about it?

Best regards,
Bryan

File: hipparchus-optim/src/test/java/org/hipparchus/optim/nonlinear/vector/leastsquares/SequentialGaussNewtonOptimizerWithCholeskyTest.java [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+        CircleVectorial circle = new CircleVectorial();
+        circle.addPoint( 30.0,  68.0);
+        circle.addPoint( 50.0,  -6.0);
+        circle.addPoint(110.0, -20.0);
+        circle.addPoint( 35.0,  15.0);
+        circle.addPoint( 45.0,  97.0);
+
+        LeastSquaresProblem lsp = builder(circle)
+                .checkerPair(new SimpleVectorValueChecker(1e-30, 1e-30))
+                .maxIterations(Integer.MAX_VALUE)
```
Comment (maisonobe) [inline hipparchus-optim/src/test/java/org/hipparchus/optim/nonlinear/vector/leastsquares/SequentialGaussNewtonOptimizerWithCholeskyTest.java:79]: This test generates an infinite loop.
It seems related to commit 653fd4c12cac8c028 as `evaluationCounter` has been removed.
As only this counter was set to a small value and only `iterationCounter` remains and is set to `Integer.MAX_VALUE`, it cannot be exceeded and the test loops. The same happen also for other decomposition algorithms.

I will therefore put `evaluationCounter` back before merging the PR.

File: hipparchus-optim/src/test/java/org/hipparchus/optim/nonlinear/vector/leastsquares/SequentialGaussNewtonOptimizerWithQRTest.java [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+ * <p>Some of the unit tests are re-implementations of the MINPACK <a
+ * href="http://www.netlib.org/minpack/ex/file17">file17</a> and <a
+ * href="http://www.netlib.org/minpack/ex/file22">file22</a> test files.
+ * The redistribution policy for MINPACK is available <a
+ * href="http://www.netlib.org/minpack/disclaimer">here</a>/
+ *
+ */
+public class SequentialGaussNewtonOptimizerWithQRTest
+    extends AbstractSequentialLeastSquaresOptimizerAbstractTest {
+
```
Comment (maisonobe) [inline hipparchus-optim/src/test/java/org/hipparchus/optim/nonlinear/vector/leastsquares/SequentialGaussNewtonOptimizerWithQRTest.java:41]: I don't see the difference between this test and SequentialGaussNewtonOptimizerWithQRNormalTest,
apart from some methods moved around and the expected exception check differently.
Shouldn't SequentialGaussNewtonOptimizerWithQRNormalTest be removed?

**Comment by maisonobe** [comment]
I have reviewed more thoroughly the PR and made some changes in a new branch bayardj-sequential-gauss-newton
in the official repository.
The main change was to allow using a null previous evaluation. This seemed useful to me when starting the process from scratch.
I have also added a test that really exercises sequential feature as the existing tests in fact started with a dummy CheckerEvaluation filled with zeroes and performed only one call with all measurements provided at once. This in fact ended up with an optimum that considered having twice the number of measurements (the dummy ones at zero followed by the real ones.

One test Hahn1 which throwed an exception before my changes now completes but the results are very bad. I think however this test is a difficult one, it is seen as singular with other optimizers.

Could you review these changes and tell me if you agree with them?

**Comment by maisonobe** [comment]
I found the reason why `SequentialGaussNewtonOptimizerWithSVDTest.testHahn1` failed with an exception previously but not anymore.
The test in the PR was wrong. It used `@Test(expected=MathIllegalStateException.class)` which only checks
the type of the exception, but in fact the setup in `super.testHahn1()` used the wrong dimensions for the problem when setting up the initial `CheckEvaluation` instance (it used 1 instead of 7). So the test failed earlier than expected due to a `LocalizedCoreFormats.DIMENSIONS_MISMATCH` exception. In Hipparchus tests, it is better to not use  `@Test(expected=SomeException.class)`, but rather to catch the exception in the test and to compare `e.getSpecifier()`,
and sometimes even `e.getParts()` against an expected reference.
When using the correct dimension in the setting, the same behaviour occurs in the PR and in my changed branch: optimization completes and results are numerically very far for expected results.

**Comment by maisonobe** [comment]
I have added a few more changes and merged the PR.
The changes are:
  - allow user to build or not the normal equations
  - allow user to start with only a priori state and covariance (thanks to Evan for the hint!)

**Pull request merged**